### ISSUE TITLE
undefined reference to `pcap_debug'

### ISSUE THREAD

**Issue by mudongliang** [the-tcpdump-group/libpcap#489]
I tried to compile tcpdump 3.8.1 from source code in Ubuntu (for my experiment). When I `make` to generate `tcpdump`, error happened - "undefined reference to `pcap_debug'".
I guessed the libpcap installed from the official software repos is not suitable. So I downloaded libpcap 0.8.1(corresponding to tcpdump 3.8.1) in the same directory so that tcpdump could use it. However, the same error ...
Then I tried two versions : 0.7.2, 1.6.2. No one worked.
How to solve this problem?

**Comment by infrastation** [comment]
tcpdump version 3.8.1 is not supported because it dates back to December 2003. Do you see this error with the latest version of tcpdump and libpcap?

**Comment by mcr** [comment]
慕冬亮 <notifications@github.com> wrote:
    > I tried to compile tcpdump 3.8.1 from source code in Ubuntu (for my
    > experiment). When I make to generate tcpdump, error happened -
    > "undefined reference to `pcap_debug'".  I guessed the libpcap installed
    > from the official software repos is not suitable. So I downloaded
    > libpcap 0.8.1(corresponding to tcpdump 3.8.1) in the same directory so

Assuming that you have some legit reason to use 13 year old code,
I would suspect that the ./configure for tcpdump did not find the libpcap
0.8.1 compiled version.  Did you run ./configure again after compiling
libpcap?

Did Ubuntu even exist in 2003?

-- 
]               Never tell me the odds!                 | ipv6 mesh networks [ 
]   Michael Richardson, Sandelman Software Works        | network architect  [ 
]     mcr@sandelman.ca  http://www.sandelman.ca/        |   ruby on rails    [

**Comment by mudongliang** [comment]
This is the version of libpcap and tcpdump:

    libpcap-0.8.1   tcpdump-3.8.1

After I compiled the libpcap-0.8.1 successfully, I compiled tcpdump with one of output in the following:

    gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c version.c

The same error happened.

After I changed to another os, CentOS and tried to compile, same error happened.

**Comment by guyharris** [comment]
Try running `make distclean` in the tcpdump source directory, and then run the command

    ./configure | egrep pcap_debug

and paste the results here.  If it says

    checking whether pcap_debug is defined by libpcap... yes

please open the file `config.log` in an editor, look for the line containing the string

    checking whether pcap_debug is defined by libpcap

and paste everything from that line to the line containing the string

    checking for pcap_set_optimizer_debug

along with the output of the configure script.

**Comment by mudongliang** [comment]
After I `make distclean`, I ran the command - `./configure | egrep pcap_debug`. Outputs are in the following:

    $ ./configure | egrep pcap_debug
    configure: warning: The SMB printer may have exploitable buffer overflows!!!
    checking whether pcap_debug is defined by libpcap... yes

However, I didn't find `pcap_set_optimizer_debug`, so I greped all the pcap from config.log

```
$ grep pcap config.log
configure:3642: checking for local pcap library
configure:3750: checking for pcap_list_datalinks
configure:3778: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:3802: checking for pcap_set_datalink
configure:3830: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:3853: checking for pcap_datalink_name_to_val
configure:3881: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:3901: checking for pcap_datalink_val_to_description
configure:3929: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:3962: checking for pcap_breakloop
configure:3990: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4042: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4067: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4092: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4147: checking for pcap_findalldevs
configure:4175: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4147: checking for pcap_dump_flush
configure:4175: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4147: checking for pcap_lib_version
configure:4175: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4239: checking whether pcap_debug is defined by libpcap
configure:4256: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4346: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4904: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing -L/usr/lib  conftest.c -lcrypto  ./../libpcap-0.8.1/libpcap.a  1>&5
configure:4947: gcc -E  -I$(srcdir)/missing -I./../libpcap-0.8.1  -I$(srcdir)/missing -I/usr/include conftest.c >/dev/null 2>conftest.out
```

**Comment by guyharris** [comment]
So what are the next 200 lines after

    configure:4239: checking whether pcap_debug is defined by libpcap

in the `config.log` file?

**Comment by mudongliang** [comment]
```
333 configure:4239: checking whether pcap_debug is defined by libpcap
334 configure:4256: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
335 configure:4318: checking for bpf_dump
336 configure:4346: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
337 configure:4403: checking for int8_t using gcc
338 configure:4421: gcc -c -DINET6 -g -O2  -I$(srcdir)/missing conftest.c 1>&5
339 configure:4441: checking for u_int8_t using gcc
340 configure:4459: gcc -c -DINET6 -g -O2  -I$(srcdir)/missing conftest.c 1>&5
341 configure:4479: checking for int16_t using gcc
342 configure:4497: gcc -c -DINET6 -g -O2  -I$(srcdir)/missing conftest.c 1>&5
343 configure:4517: checking for u_int16_t using gcc
344 configure:4535: gcc -c -DINET6 -g -O2  -I$(srcdir)/missing conftest.c 1>&5
345 configure:4555: checking for int32_t using gcc
346 configure:4573: gcc -c -DINET6 -g -O2  -I$(srcdir)/missing conftest.c 1>&5
347 configure:4593: checking for u_int32_t using gcc
348 configure:4611: gcc -c -DINET6 -g -O2  -I$(srcdir)/missing conftest.c 1>&5
349 configure:4671: checking if sockaddr struct has sa_len member
350 configure:4685: gcc -c -DINET6 -g -O2  -I$(srcdir)/missing conftest.c 1>&5
351 configure: In function 'main':
352 configure:4681: error: 'struct sockaddr' has no member named 'sa_len'
353 configure: failed program was:
354 #line 4676 "configure"
355 #include "confdefs.h"
356 
357 #       include <sys/types.h>
358 #       include <sys/socket.h>
359 int main() {
360 u_int i = sizeof(((struct sockaddr *)0)->sa_len)
361 ; return 0; }
362 configure:4706: checking if unaligned accesses fail
363 configure:4802: checking for h_errno
364 configure:4816: gcc -c -DINET6 -g -O2  -I$(srcdir)/missing conftest.c 1>&5
365 configure:4844: checking for SSLeay
366 configure:4885: checking for des_cbc_encrypt in -lcrypto
367 configure:4904: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing -L/usr/lib  conftest.c -lcrypto  ./../libpcap-0.8.1/libpcap.a  1>&5
368 /tmp/ccm4oCyP.o: In function `main':
369 /home/core/tcpdump/tcpdump-3.8.1/configure:4900: undefined reference to `des_cbc_encrypt'
370 collect2: ld returned 1 exit status
371 configure: failed program was:
372 #line 4893 "configure"
373 #include "confdefs.h"
374 /* Override any gcc2 internal prototype to avoid an error.  */
375 /* We use char because int might match the return type of a gcc2
376     builtin and then its argument prototype would still apply.  */
377 char des_cbc_encrypt();
378 
379 int main() {
380 des_cbc_encrypt()
381 ; return 0; }
382 configure:4937: checking for openssl/evp.h
383 configure:4947: gcc -E  -I$(srcdir)/missing -I./../libpcap-0.8.1  -I$(srcdir)/missing -I/usr/include conftest.c >/dev/null 2>conftest.out
384 configure:4997: checking for a BSD compatible install
```

**Comment by guyharris** [comment]
333 configure:4239: checking whether pcap_debug is defined by libpcap
    334 configure:4256: gcc -o conftest -DINET6 -g -O2  -I$(srcdir)/missing  conftest.c ./../libpcap-0.8.1/libpcap.a  1>&5
    335 configure:4318: checking for bpf_dump

Then either 1) `pcap_debug` *is* defined by `./../libpcap-0.8.1/libpcap.a` or 2) the linker is behaving badly.

What are the last 50 lines of the output of the build when you try to build tcpdump and it fails with an undefined reference `pcap_debug`?

**Comment by mudongliang** [comment]
I have used `nm libpcap.a | grep pcap_debug` to find the missing symbol. But I didn't search it.

And the following is the output of `make`
```
sed -e 's/.*/char version[] = "&";/' ./VERSION > version.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c version.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-ip6.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-ip6opts.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-mobility.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-ripng.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-icmp6.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-frag6.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-rt6.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-ospf6.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-dhcp6.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./print-smb.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -c ./smbutil.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -o strlcat.o -c ./missing/strlcat.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -o strlcpy.o -c ./missing/strlcpy.c
gcc -O2 -DHAVE_CONFIG_H  -D_U_="" -I. -I./../libpcap-0.8.1  -I./missing -I/usr/include -L/usr/lib  -o tcpdump addrtoname.o gmpls.o gmt2local.o machdep.o parsenfsfh.o print-802_11.o print-ah.o print-arcnet.o print-aodv.o print-arp.o print-ascii.o print-atalk.o print-atm.o print-beep.o print-bfd.o print-bgp.o print-bootp.o print-cdp.o print-chdlc.o print-cip.o print-cnfp.o print-decnet.o print-domain.o print-dvmrp.o print-enc.o print-egp.o print-esp.o print-ether.o print-fddi.o print-fr.o print-gre.o print-hsrp.o print-icmp.o print-igmp.o print-igrp.o print-ip.o print-ipcomp.o print-ipfc.o print-ipx.o print-isakmp.o print-isoclns.o print-krb.o print-l2tp.o print-lane.o print-ldp.o print-llc.o print-lwres.o print-mobile.o print-mpls.o print-msdp.o print-nfs.o print-ntp.o print-null.o print-ospf.o print-pflog.o print-pim.o print-ppp.o print-pppoe.o print-pptp.o print-radius.o print-raw.o print-rip.o print-rsvp.o print-rx.o print-sctp.o print-sl.o print-sll.o print-snmp.o print-stp.o print-sunatm.o print-sunrpc.o print-tcp.o print-telnet.o print-tftp.o print-timed.o print-token.o print-udp.o print-vjc.o print-vrrp.o print-wb.o print-zephyr.o setsignal.o tcpdump.o util.o version.o print-ip6.o print-ip6opts.o print-mobility.o print-ripng.o print-icmp6.o print-frag6.o print-rt6.o print-ospf6.o print-dhcp6.o print-smb.o smbutil.o strlcat.o strlcpy.o ./../libpcap-0.8.1/libpcap.a 
tcpdump.o: In function `main':
tcpdump.c:(.text+0x6f2): undefined reference to `pcap_debug'
collect2: ld returned 1 exit status
make: *** [tcpdump] Error 1
```

**Comment by guyharris** [comment]
That's a tcpdump bug fixed by [this patch](https://sourceforge.net/p/tcpdump/patches/43/).  That fix isn't in tcpdump 3.8.1, or 3.8.2, or 3.8.3; the first tcpdump release with the fix is tcpdump 3.9.1.

So you will either have to use tcpdump 3.9.1 or apply [this change](https://github.com/the-tcpdump-group/tcpdump/commit/0b18136512b0ec413821e5210ab5f343f622ee98l) to whatever earlier version of tcpdump you are trying to build.
